### PR TITLE
Semantic Text - Use offsets from chunked inference response

### PR DESCRIPTION
**Pull request by Mikep86**
Use the offsets from the `ChunkedInference` response when indexing semantic text field values. When using the inference metadata fields format, we store the offsets directly. When using the legacy format, we use the offsets to build the chunk text and store that.

This change should allow us up to remove `ChunkedInference.Chunk#matchedText` in a follow-up, which should help with OOMs when chunking large documents.

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java [hunk 1]
```diff
@@ -425,15 +425,16 @@ public static Chunk toSemanticTextFieldChunk(
...
-            chunk.bytesReference()
-        );
+        String text = null;
+        int startOffset = -1;
+        int endOffset = -1;
+        if (useInferenceMetadataFieldsFormat) {
+            startOffset = chunk.textOffset().start() + offsetAdjustment;
+            endOffset = chunk.textOffset().end() + offsetAdjustment;
+        } else {
+            text = input.substring(chunk.textOffset().start(), chunk.textOffset().end());
```
Comment (jimczi) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:435]: Can we also don't keep the `input` around if `useInferenceMetadataFieldsFormat` is true? I mean we can save some memory in `ShardBulkInferenceAction` by not keeping the input in such case.

**Review by jimczi — APPROVED** [review]
LGTM
I left one small comment, feel free to address it here or we can have a follow up.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java (hunk 1 above)
Comment (Mikep86) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:435]: Sounds like a good optimization for a follow-up IMO. We need to keep the input in at least `FieldInferenceRequest` for actually performing inference, but we could conditionally set it in `FieldInferenceResponse`.

**Review by carlosdelest — APPROVED** [review]

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java (hunk 1 above)
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:435]: sure

**Review by kderusso — APPROVED** [review]
LGTM!

**Pull request merged**